### DISCUSSION TITLE
Working with data from external api

### DISCUSSION THREAD

**Discussion by luctus** [bridgetownrb/bridgetown#454] [Q&A] [answered]
Hello guys,

I'm sorry to write another post here, but I really can't find much information on internet yet... I'd really appreciate it if you can give me a little guidance here.

## Context
I'm using bridgetown edge, based on [this article](https://edge.bridgetownrb.com/docs/plugins/external-apis). I created a collection called `shops` and I'm loading the list of shops from an external api. I want to have a page where you list all the shops and then one page per shop. I'm using ERB.

## Setup
My `bridgetown.config.yml` looks like this:
```yaml
...
pagination:
  enabled: true
  per_page: 10
template_engine: erb
collections:
  shops:
    output: true
    permalink: /shops/:slug
    sort_by: name
    sort_direction: ascending
    
```

I created this builder:
```rb
class Builders::LoadOffersFromApi < SiteBuilder
  def build
    get "https://my-super-endpoint.com/shops.json" do |data|  
      data.each_with_index do |brand, index|
        add_resource :shops, "#{brand[:slug]}.erb" do
          layout :shops
          id brand[:id]
          name brand[:name]
          slug brand[:slug]
          description brand[:description]
          position index
        end
      end
  end
end
```

Then, I have the template for the list of the shops, at `src/shops.erb` (this is a simplified version):
```erb
---
layout: default
pagination:
  collection: shops
---

<h1>All shops</h1>
<% paginator.resources.each do |shop| %>
  <a href="<%= shop.relative_url %>">
    <%= shop.data.name %>
  </a>
<% end %>
```

And I also created a layout for rendering the pages for each shop, at `src/_layouts/shops.erb` (this is a simplified version):
```erb
---
layout: default
---

<h1>Shop: <%=  page.data.name %></h1>
```
I also created the layout called `default`, in `erb` format that basically contains the `<%= yield %>`.

## Problems

- I can see the list of shops at `/shops`, but they are not ordered by name
- I can load the page of a shop, like `/shops/amazon`, but I see the html code rendered in the browser, not interpreted

That's it. I'd really appreciate any guidance.

Thanks in advance!

**Comment by jaredcwhite** [comment, marked as answer]
Looks pretty solid at first glance! I think regarding the raw html issue you'd need to change the permalink to either `/shops/:slug/` or `/shops/:slug.*` (depending on if you want the URL to end with a slash or not)

Not sure why the ordering isn't working tho…

**Reply by luctus** [reply]
thank you @jaredcwhite! The trailing slash did all the magic :)

I'll continue debuging to understand what's going on with the ordering thing.

Have a great day!

**Reply by luctus** [reply]
Hey @jaredcwhite,

I haven't been able to fix the ordering issue... I found that the paginator is using a `config[:sort_field]` too... quite confusing.

https://github.com/bridgetownrb/bridgetown/blob/4182c8de1674fbfa2e4d0d31370bb9c8828633dc/bridgetown-paginate/lib/bridgetown-paginate/pagination_model.rb#L170

Do you know how could I debug it in a better way?

Thanks in advance.

**Reply by luctus** [reply]
aha! having this in the template:

```erb
---
layout: default
pagination:
  collection: shops
  per_page: 20
  sort_field: global_position
  sort_reverse: false
---
```

Did the trick!

**Reply by jaredcwhite** [reply]
@luctus So if you add `sort_field: name` nested under the `pagination:` key in your `shops.erb` file, that might do the trick…but by default the paginator shouldn't be resorting anything so still a bit of a mystery. But if that does fix it for now, I'll file an issue to look into that further.

**Reply by jaredcwhite** [reply]
Ah, posted my comment before seeing your latest. OK! Well making progress here. 😅 Still feels like a bug may be lurking here so I'll poke around some more. But glad you got that working.

**Reply by luctus** [reply]
I have to say that I'm loving it so far. Thanks a lot for all your help.
